hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1092,7 +1092,7 @@ protected boolean isSubclassTableIndicatedByTreatAsDeclarations(
 	private String[] getSubclassNameClosureBySubclassTable(int subclassTableNumber) {
 		final int index = subclassTableNumber - getTableSpan();
 
-		if ( index > subclassNamesBySubclassTable.length ) {
+		if ( index >= subclassNamesBySubclassTable.length ) {
 			throw new IllegalArgumentException(
 					"Given subclass table number is outside expected range [" + subclassNamesBySubclassTable.length
 							+ "] as defined by subclassTableNameClosure/subclassClosure"
```

Comment:
The upper limit should be corrected in the error message to be ```subclassNamesBySubclassTable - 1```.

Reply:
Changed as suggested